vars/vars_tracker.go

```diff
@@ -82,29 +84,11 @@ func (t *credVarsTracker) IterateInterpolatedCreds(iter CredVarsTrackerIterator)
 }
 
 func (t *credVarsTracker) Enabled() bool {
-	return true
-}
-
-// DummyCredVarsTracker do nothing,
-
-type dummyCredVarsTracker struct {
-	credVars Variables
-}
-
-func (t dummyCredVarsTracker) Get(varDef VariableDefinition) (interface{}, bool, error) {
-	return t.credVars.Get(varDef)
-}
-
-func (t dummyCredVarsTracker) List() ([]VariableDefinition, error) {
-	return t.credVars.List()
-}
-
-func (t dummyCredVarsTracker) IterateInterpolatedCreds(iter CredVarsTrackerIterator) {
-	// do nothing
+	return t.enabled
 }
 
-func (t dummyCredVarsTracker) Enabled() bool {
-	return false
+func (t *credVarsTracker) AddVar(v Variables) {
+	t.credVars = NewMultiVars([]Variables{v, t.credVars})
```

Comment:
An alternative could be to instantiate `credVars` with a [`vars.StaticVariables`](https://github.com/concourse/concourse/blob/master/vars/static_vars.go#L7) embedded that we then mutate in this method.

Something like this:

```go
func NewCredVarsTracker(credVars Variables, on bool) CredVarsTracker {
  localVars := StaticVariables{}

  return &credVarsTracker{
    localVars: localVars,

    // note: I can never remember the proper order here...
    credVars: NewMultiVars([]Variables{localVars, credVars}),

    // ...
  }
}

func (t *credVarsTracker) SetLocalVar(name string, val interface{}) {
  t.localVars[name] = val
}
```

Splitting hairs, but this would result in less allocation and traversal through nested `NewMultiVars`. And maybe if we figure out a distinct syntax for build-local vars we wouldn't even need to mess with `credVars`. :thinking:

Reply:
Comparing with secret redaction, I guess cost of nested `MultiVars` is tiny. I think let's only address this issue once you work out the final RFC. Maybe in the RFC, you will also address if local var should be redacted, maybe have an attribute for `var` step to indicate if a local var should be redacted.

Reply:
I only meant to suggest a refactor here because it seems kind of confusing to keep chaining `StaticVariables` that each contain only one entry. I wasn't really thinking about secret redacting. :confused: I guess we can re-assess this when we implement local vars, in which case having a distinct separation would be more important so you can't accidentally access a local var with normal var syntax. Not a big deal right now. :+1: 

I can see why might want an option to toggle redaction; good point. The var step could be used to read an innocuous value like a dynamically generated Docker tag, in which case you probably *want* that to be printed in your build for clarity. For safety reasons I'll probably have them redact by default though in the RFC. We can keep talking about this point there - I'll try to have something out today.